pkg/restore/pipeline_items.go

```diff
@@ -0,0 +1,338 @@
+const (
+	defaultBatcherOutputChannelSize = 1024
+)
+
+// CreatedTable is a table created on restore process,
+// but not yet filled with data.
+type CreatedTable struct {
+	RewriteRule *RewriteRules
+	Table       *model.TableInfo
+	OldTable    *utils.Table
+}
+
+// TableWithRange is a CreatedTable that has been bind to some of key ranges.
+type TableWithRange struct {
+	CreatedTable
+
+	Range []rtree.Range
+}
+
+// Batcher collects ranges to restore and send batching split/ingest request.
+type Batcher struct {
+	cachedTables   []TableWithRange
+	cachedTablesMu *sync.Mutex
+	rewriteRules   *RewriteRules
+
+	// joiner is for joining the background batch sender.
+	joiner             chan<- struct{}
+	sendErr            chan<- error
+	outCh              chan<- CreatedTable
+	sender             BatchSender
+	batchSizeThreshold int
+	size               int32
+}
+
+// SetThreshold sets the threshold that how big the batch size reaching need to send batch.
+// note this function isn't goroutine safe yet,
+// just set threshold before anything starts(e.g. EnableAutoCommit), please.
+func (b *Batcher) SetThreshold(newThreshold int) {
+	b.batchSizeThreshold = newThreshold
+}
+
+// Exhaust drains all remaining errors in the channel, into a slice of errors.
+func Exhaust(ec <-chan error) []error {
+	out := make([]error, 0, len(ec))
+	for {
+		select {
+		case err := <-ec:
+			out = append(out, err)
+		default:
+			// errCh will *never* closed(ya see, it has multi sender-part),
+			// so we just consume the current backlog of this cannel, then return.
+			return out
+		}
+	}
+}
+
+// Len calculate the current size of this batcher.
+func (b *Batcher) Len() int {
+	return int(atomic.LoadInt32(&b.size))
+}
+
+// BatchSender is the abstract of how the batcher send a batch.
+type BatchSender interface {
+	// RestoreBatch will send the restore request.
+	RestoreBatch(ctx context.Context, ranges []rtree.Range, tbs []CreatedTable) error
+	Close()
+}
+
+type tikvSender struct {
+	client         *Client
+	updateCh       glue.Progress
+	rejectStoreMap map[uint64]bool
+}
+
+// NewTiKVSender make a sender that send restore requests to TiKV.
+func NewTiKVSender(ctx context.Context, cli *Client, updateCh glue.Progress) (BatchSender, error) {
+	tiflashStores, err := conn.GetAllTiKVStores(ctx, cli.GetPDClient(), conn.TiFlashOnly)
+	if err != nil {
+		// After TiFlash support restore, we can remove this panic.
+		// The origin of this panic is at RunRestore, and its semantic is nearing panic, don't worry about it.
+		log.Error("failed to get and remove TiFlash replicas", zap.Error(errors.Trace(err)))
+		return nil, err
+	}
+	rejectStoreMap := make(map[uint64]bool)
+	for _, store := range tiflashStores {
+		rejectStoreMap[store.GetId()] = true
+	}
+
+	return &tikvSender{
+		client:         cli,
+		updateCh:       updateCh,
+		rejectStoreMap: rejectStoreMap,
+	}, nil
+}
+
+func (b *tikvSender) RestoreBatch(ctx context.Context, ranges []rtree.Range, tbs []CreatedTable) error {
```

Comment:
This is buggy since we just send 'empty' tables to `RestoreBatch`. If partial range of table is sent here, its rewrite rules won't be proceed.

`tbs` here seems redundant, maybe replace it with `rewriteRules` will be better, and we need test cases.